### ISSUE TITLE
Logsumexp is not numerically stable

### ISSUE THREAD

**Issue by ryotatomioka** [onnx/onnx-mlir#365]
[Rewriting](https://github.com/onnx/onnx-mlir/blob/master/src/Transform/ONNX/Decompose.td#L46) of logsumexp in terms of simpler onnx ops need to take into account the numerical stability.

To reproduce the issue,
```python
import torch
import torch.nn as nn

class MyModule(nn.Module):
    def forward(self, x):
        return torch.logsumexp(x, dim=1)

model = MyModule()
x = torch.randn(3, 4)
torch.onnx.export(model, x, "logsumexp.onnx")

import sys
sys.path = ['test/backend'] + sys.path
import test

import onnx
onnx_model = onnx.load("logsumexp.onnx")
session = test.DummyBackend.prepare(onnx_model)
import numpy as np
x = np.random.normal(0, 1, (3, 4)).astype(np.float32) + 100.0
out = session.run(x)[0]
expected = model(torch.tensor(x)).numpy()
print(f"out={out}")
print(f"logsumexp(x)={expected}")
np.testing.assert_allclose(out, expected, rtol=1e-6)
```

The result was
```
out=[inf inf inf]
logsumexp(x)=[102.05343  101.04576  101.345764]
```
and the assertion fails. Note that if I comment out `+ 100.0`, then the test passes.

To make logsumexp numerically stable, we need to implement it as `logsumexp(x, dim=axis) = log(sum(exp(x - max), dim=axis)) + max` where `max=reduce_max(x, dim=axis, keepdims=True)`.

Similarly it would be better to implement `logsoftmax(x)` as `logsoftmax(x, dim=axis) = x - logsumexp(x, dim=axis)` to avoid computing the logarithm of a very small number.

**Comment by tungld** [comment]
Hi @ryotatomioka,

Thank you for reporting the issues!

I fixed ONNXReduceLogSumExp with this PR #376. 

For ONNXLogSoftmax, it requires a little work since the way it uses `axis` is different from the one ONNXReduceLogSumExp does.
